src/adaXT/random_forest/RandomForest.py

```diff
@@ -0,0 +1,355 @@
+            DecisionTree(
+                tree_type=self.forrest_type,
+                criteria=self.criterion,
+                max_depth=self.max_depth,
+                impurity_tol=self.impurity_tol,
+                min_samples_split=self.min_samples_split,
+                min_samples_leaf=self.min_samples_leaf,
+                min_improvement=self.min_improvement,
+                max_features=self.max_features,
+                splitter=self.splitter) for _ in range(
+                self.n_estimators)]
+
+        # Fit trees
+        self.__build_trees()
+
+        # Register that the forrest was succesfully fitted
+        self.forest_fitted = True
+
+        return self
+
+    # Function that returns an ndarray of random ints used as the
+    # sample_indices
+    def __get_sample_indices(self):
+        if self.bootstrap:
+            return np.random.randint(
+                low=0, high=self.n_obs, size=self.max_samples)
+        else:
+            return None
+
+    def predict(self, X: np.ndarray):
+        """
+        Predicts a y-value for given X values using the trees of the forrest.
+
+        Parameters
+        ----------
+        X : np.ndarray
+            (N, M) numpy array with features to predict
+
+        Returns
+        -------
+        np.ndarray
+            (N) numpy array with the prediction
+        """
+        if not self.forest_fitted:
+            raise AttributeError(
+                "The forrest has not been fitted before trying to call predict")
+
+        self.predict_values = SharedNumpyArray(X)
+
+        # Predict using all the trees, each column is the predictions from one
+        # tree
+        tree_predictions = self.__predict_trees()
+        self.predict_values.unlink()
+
+        if self.forrest_type == "Regression":
+            # Return the mean answer from all trees for each row
+            return np.mean(tree_predictions, axis=1)
+
+        elif self.forrest_type == "Classification":
+            # Return the element most voted for by trees for each row
+            # QUESTION: how to handle equal majority vote?
+            return np.apply_along_axis(
+                self.most_frequent_element, 1, tree_predictions)
+
+    # Function used to find the most frequent element of an array
+    def most_frequent_element(self, arr):
+        values, counts = np.unique(arr, return_counts=True)
+        return values[np.argmax(counts)]
+
+    def predict_proba(self, X: np.ndarray):
+        """
+        Predicts a probability for each response for given X values using the trees of the forrest
+
+        Parameters
+        ----------
+        X : np.ndarray
+            (N, M) numpy array with features to predict
+
+        Returns
+        -------
+        np.ndarray
+            Returns an ndarray with the probabilities for each class per observation in X. The order of the classes corresponds to that in the attribute classes
+        """
+        if not self.forest_fitted:
```

Comment:
Again we should harmonize this with how we deal with predict_proba in the decision tree.

Reply:
I think this is better solved in the PR that implements Predict and LeafBuilder classes.